File: packages/python/src/openskp/vff.py
"""VFF/ZIP container handler for SketchUp files.

SketchUp ``.skp`` files are VFF containers that embed a ZIP archive at a
known offset.  The ZIP contains at minimum ``model.dat`` (the TLV binary
data) and optionally material texture/XML files.

This module validates the VFF header, locates the embedded ZIP, and
extracts the relevant entries.
"""

from __future__ import annotations

import io
import zipfile
from typing import Any, Dict


# ── VFF magic bytes ───────────────────────────────────────────────────────

_VFF_MAGIC: bytes = b"\xFF\xFE\xFF\x0E"
"""First four bytes expected in a valid SketchUp VFF container."""

_ZIP_LOCAL_HEADER: bytes = b"PK\x03\x04"
"""ZIP local-file header signature used for scanning."""


# ── Internal helpers ─────────────────────────────────────────────────────


def _find_zip_offset(data: bytes) -> int:
    """Locate the start of the embedded ZIP archive inside *data*.

    The function scans for a ``PK\\x03\\x04`` ZIP local-file header.

    Args:
        data: Raw file bytes.

    Returns:
        Byte offset of the first ZIP header.

    Raises:
        ValueError: If no ZIP signature is found.
    """
    offset = data.find(_ZIP_LOCAL_HEADER)
    if offset < 0:
        raise ValueError("No embedded ZIP archive found in the file")
    return offset


def _read_version(data: bytes) -> str:
    """Extract the SketchUp version string from the VFF header.

    The version is a UTF-16LE encoded string between FF FE FF markers
    in the file header, enclosed in braces (e.g. ``"{25.0.575}"``).

    Args:
        data: Raw file bytes.

    Returns:
        Version string (e.g. ``"{25.0.575}"``), or ``"unknown"`` if
        not found.
    """
    if len(data) < 16:
        return "unknown"

    # Find second FF FE FF marker after the initial one at offset 0
    second_marker = data.find(b'\xFF\xFE\xFF', 4)
    if second_marker > 0:
        ver_start = second_marker + 4
        ver_bytes = data[ver_start:ver_start + 200]
        ver_text = ver_bytes.decode('utf-16-le', errors='ignore')
        brace_start = ver_text.find('{')
        brace_end = ver_text.find('}')
        if brace_start >= 0 and brace_end > brace_start:
            return ver_text[brace_start:brace_end + 1]

    return "unknown"


# ── Public API ────────────────────────────────────────────────────────────


def validate_header(data: bytes) -> bool:
    """Check whether *data* begins with the VFF magic bytes.

    Args:
        data: Raw file bytes (at least 4 bytes).

    Returns:
        ``True`` if the header matches a SketchUp VFF container.
    """
    return data[:4] == _VFF_MAGIC


def extract_skp_contents(data: bytes) -> Dict[str, Any]:
    """Extract ``model.dat`` and material files from a ``.skp`` byte buffer.

    Args:
        data: Complete ``.skp`` file contents.

    Returns:
        A dict with the following keys:

        * ``"version"`` — integer file-format version.
        * ``"model_data"`` — raw bytes of the ``model.dat`` entry.
        * ``"material_files"`` — mapping of filename → bytes for every
          material-related entry found in the ZIP.

    Raises:
        ValueError: If *data* is not a recognised SketchUp file or does
            not contain a ``model.dat`` entry.
    """
    # Allow both VFF-wrapped and bare ZIP (some exporters omit the header)
    if not validate_header(data):
        if _ZIP_LOCAL_HEADER not in data[:64]:
            raise ValueError("Not a valid SketchUp (.skp) file")

    version = _read_version(data)
    zip_offset = _find_zip_offset(data)

    zip_bytes = data[zip_offset:]
    zf = zipfile.ZipFile(io.BytesIO(zip_bytes))

    model_data: bytes = b""
    material_files: Dict[str, bytes] = {}

    for entry in zf.namelist():
        lower = entry.lower()
        if lower == "model.dat" or lower.endswith("/model.dat"):
            model_data = zf.read(entry)
        elif (
            lower.endswith(".xml")
            or lower.endswith(".png")
            or lower.endswith(".jpg")
            or lower.endswith(".jpeg")
            or "material" in lower
        ):
            material_files[entry] = zf.read(entry)

    zf.close()

    if not model_data:
        raise ValueError(
            "ZIP archive found but does not contain a model.dat entry"
        )

    return {
        "version": version,
        "model_data": model_data,
        "material_files": material_files,
    }

File: packages/dotnet/OpenSkp/Vff.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Text;

namespace OpenSkp
{
    /// <summary>Container-level helpers for the modern (2021+) VFF/ZIP .skp
    /// format: header validation, version-string extraction, and locating the
    /// embedded ZIP payload. Ported from Python's _core.full_parse() steps
    /// that run before TLV parsing begins.</summary>
    internal static class Vff
    {
        private static readonly byte[] HeaderMagic = { 0xFF, 0xFE, 0xFF, 0x0E };
        private static readonly byte[] SecondMarker = { 0xFF, 0xFE, 0xFF };
        private static readonly byte[] ZipMagic = { (byte)'P', (byte)'K', 0x03, 0x04 };

        public static bool HasValidHeader(byte[] data)
        {
            if (data.Length < 4) return false;
            for (int i = 0; i < 4; i++)
            {
                if (data[i] != HeaderMagic[i]) return false;
            }
            return true;
        }

        private static int FindBytes(byte[] data, byte[] pattern, int start)
        {
            if (pattern.Length == 0 || start < 0) return -1;
            int limit = data.Length - pattern.Length;
            for (int i = Math.Max(start, 0); i <= limit; i++)
            {
                bool match = true;
                for (int j = 0; j < pattern.Length; j++)
                {
                    if (data[i + j] != pattern[j]) { match = false; break; }
                }
                if (match) return i;
            }
            return -1;
        }

        /// <summary>Extract the "{17.0.18899}"-style version string from the
        /// UTF-16LE-encoded header region, or "unknown" if not found.</summary>
        public static string ExtractVersion(byte[] header)
        {
            int secondMarker = FindBytes(header, SecondMarker, 4);
            if (secondMarker <= 0) return "unknown";

            int verStart = secondMarker + 4;
            if (verStart >= header.Length) return "unknown";

            string verText;
            try
            {
                verText = Encoding.Unicode.GetString(header, verStart, header.Length - verStart);
            }
            catch (ArgumentException)
            {
                return "unknown";
            }

            int braceStart = verText.IndexOf('{');
            int braceEnd = verText.IndexOf('}');
            if (braceStart >= 0 && braceEnd > braceStart)
            {
                return verText.Substring(braceStart, braceEnd - braceStart + 1);
            }
            return "unknown";
        }

        /// <summary>Locate the byte offset of the embedded ZIP container
        /// ("PK\x03\x04"), searching only the first `searchLimit` bytes
        /// (matching Python's header-then-4096-byte-chunk search), or -1 if
        /// not found.</summary>
        public static int FindZipOffset(byte[] data, int searchLimit = 4096)
        {
            int limit = Math.Min(searchLimit, data.Length);
            return FindBytes(data, ZipMagic, 0, limit);
        }

        private static int FindBytes(byte[] data, byte[] pattern, int start, int end)
        {
            if (pattern.Length == 0 || start < 0) return -1;
            int limit = Math.Min(end, data.Length) - pattern.Length;
            for (int i = Math.Max(start, 0); i <= limit; i++)
            {
                bool match = true;
                for (int j = 0; j < pattern.Length; j++)
                {
                    if (data[i + j] != pattern[j]) { match = false; break; }
                }
                if (match) return i;
            }
            return -1;
        }

        public static ZipArchive OpenZip(byte[] data, int zipOffset)
        {
            var stream = new MemoryStream(data, zipOffset, data.Length - zipOffset, writable: false);
            return new ZipArchive(stream, ZipArchiveMode.Read, leaveOpen: false);
        }
    }
}
